### ISSUE TITLE
JedisConnectionFactory failed to get connection after configuration auto update

### ISSUE THREAD

**Issue by Jiabao-Sun** [spring-projects/spring-data-redis#2553]
Error Trace:
> java.lang.IllegalStateException: JedisConnectionFactory was destroyed and cannot be used anymore##  at org.springframework.util.Assert.state(Assert.java:76) ~[spring-core-5.3.23.jar!/:5.3.23]##  at org.springframework.data.redis.connection.jedis.JedisConnectionFactory.assertInitialized(JedisConnectionFactory.java:971) ~[spring-data-redis-2.7.3.jar!/:2.7.3]##  at org.springframework.data.redis.connection.jedis.JedisConnectionFactory.getConnection(JedisConnectionFactory.java:509) ~[spring-data-redis-2.7.3.jar!/:2.7.3]##  at org.springframework.data.redis.core.RedisConnectionUtils.fetchConnection(RedisConnectionUtils.java:193) ~[spring-data-redis-2.7.3.jar!/:2.7.3]##  at org.springframework.data.redis.core.RedisConnectionUtils.doGetConnection(RedisConnectionUtils.java:144)

We receive `EnvironmentChangeEvent` in the configuration hot update scenario. 
The `ConfigurationPropertiesRebinder` will destroy and reinitialize `JedisConnectionFactory`.

![image](https://user-images.githubusercontent.com/27403841/232101506-a84be603-69ba-4da5-ad28-69f1d7ea7879.png)

In destroy method we set the destroyed flag as true.
![image](https://user-images.githubusercontent.com/27403841/232104293-e1f18e30-06c3-42e9-ba76-da6470a93262.png)

But in initial method (afterPropertiesSet) we don't reset destroyed as false.
![image](https://user-images.githubusercontent.com/27403841/232104518-0e01d8ee-1240-4aed-8a8f-b95644acf0ea.png)

This will cause us to be unable to get a connection from the `JedisConnectionFactory` any more.
![image](https://user-images.githubusercontent.com/27403841/232103242-e756c909-21fa-47ab-99cc-f54b33ecd144.png)

**Comment by jxblum** [comment]
@Jiabao-Sun - This logic in Spring Data Redis may actually be by design and it could be Spring Boot's `ConfigurationPropertiesRebinder` that is actually and technically incorrect here. 

> DISCLAIMER: I say "... may actually be by design" since this predates me. I have only recently taken over as the lead of Spring Data Redis and assumed the module responsibilities. So, I am technically still learning about the reasons for our design decisions as well as the underlying nuances of Redis itself (i.e. outside of a Spring context). FYI, I am was formerly the lead of all things Spring for Apache Geode (SDG, STDG, SSDG and SBDG), which has considerable overlap with Redis in terms of behavior and functionality; both are key/value stores, for instance.  Anyway...

I observed these [assertions](https://github.com/spring-projects/spring-data-redis/blob/3.0.5/src/main/java/org/springframework/data/redis/connection/jedis/JedisConnectionFactory.java#L868-L871) (and specifically, this [assertion](https://github.com/spring-projects/spring-data-redis/blob/3.0.5/src/main/java/org/springframework/data/redis/connection/jedis/JedisConnectionFactory.java#L870)) in the Spring Data Redis codebase, which also originates in the `JedisConnectionFactory`, and is called before acquiring new connections ([Standalone](https://github.com/spring-projects/spring-data-redis/blob/3.0.5/src/main/java/org/springframework/data/redis/connection/jedis/JedisConnectionFactory.java#L448), [Clustered](https://github.com/spring-projects/spring-data-redis/blob/3.0.5/src/main/java/org/springframework/data/redis/connection/jedis/JedisConnectionFactory.java#L471) or [Sentinel](https://github.com/spring-projects/spring-data-redis/blob/3.0.5/src/main/java/org/springframework/data/redis/connection/jedis/JedisConnectionFactory.java#L797)).

In fact, it may not be "safe" it reacquire connections from a "closed" factory (Jedis or Lettuce), which I suppose would highly depend on the Redis (client) driver (again, Jedis, or Lettuce).

The real question is, why is Spring Boot's `ConfigurationPropertiesRebinder` (of all things) responsible for initializing and destroying the SD Redis, `JedisConnectionFactory`?

Because, this does not seem correct to me, or the responsibility of configuration property binding.

In most case, with most stores (that don't offer some type of failover), then the safest approach it to restart the (Spring) application, which will reinitialize the connection pool between a JVM fresh process and the Redis server(s).

**Comment by jxblum** [comment]
> NOTE: Sorry about the same assertions reference(s) that you included in your issue ticket description.

Is this `ConfigurationPropertiesRebinder` class of your own Spring application's doing?

It does not originate in either Spring Data Redis or Spring Boot. I am curious where this class comes from?

**Comment by jxblum** [comment]
Thinking further on this...

You'd be better off "refreshing" ([Javadoc](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/context/support/AbstractApplicationContext.html#refresh())) the Spring container (`ApplicationContext`) rather than trying to reinitialize a "named" (Redis) bean (such as the `JedisConnectionFactory`) in this manner, which arguably was never meat to be re-initialized in the first place, according to the current logic and design in SD Redis itself.

Of course, refreshing the Spring `ApplicationContext` refreshes the entire context and all Spring beans, which is more expensive than refreshing a single bean. But then, the SD Redis `JedisConnectionFactory` is not a trivial bean now is it.

Furthermore, you have control over declaring and constructing/initializing the `RedisConnectionFactory` in use here. 

That is even a requirement when initializing your Spring Data Redis application in the first place. See ref doc [in general](https://docs.spring.io/spring-data/redis/docs/current/reference/html/#redis:connectors), and using the `JedisConnectionFactory` in particular ([here](https://docs.spring.io/spring-data/redis/docs/current/reference/html/#redis:connectors:jedis)).

In other words, you must create a Spring managed bean of type `RedisConnectionFactory`, such as a new `JedisConnectionFactory` for your Spring application to even be able to connect to Redis servers.

In which case, you could extend the `JedisConnectionFactory` class and simply reset the `destroyed` field, reflectively (in `afterPropertiesSet()`. Though, I wouldn't even recommend this, since `destroyed` is an "internal" control variable that was not meant to be exposed in the API and set by the user (application) directly.

If you use this later approach, then you assume full responsibility of the extensions and behavioral override. This does not fall on Spring.
